crates/qpy/src/circuit_writer.rs

```diff
@@ -0,0 +1,910 @@
+                                qpy_data.annotation_handler.annotation_factories,
+                            )?);
+                        }
+                    }
+                }
+                OperationRef::Instruction(pyinst) => {
+                    let inst = pyinst.instruction.bind(py);
+                    match getattr_or_none(inst, "definition") {
+                        None => (),
+                        Some(definition) => {
+                            has_definition = true;
+                            data = serialize(&pack_circuit(
+                                &mut definition.extract()?,
+                                Some(py.None().bind(py)),
+                                false,
+                                qpy_data.version,
+                                qpy_data.annotation_handler.annotation_factories,
+                            )?);
+                        }
+                    }
+                }
+                OperationRef::Operation(pyoperation) => {
+                    let operation = pyoperation.operation.bind(py);
+                    match getattr_or_none(operation, "definition") {
+                        None => (),
+                        Some(definition) => {
+                            has_definition = true;
+                            data = serialize(&pack_circuit(
+                                &mut definition.extract()?,
+                                Some(py.None().bind(py)),
+                                false,
+                                qpy_data.version,
+                                qpy_data.annotation_handler.annotation_factories,
+                            )?);
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
+        let num_qubits = operation.num_qubits();
+        let num_clbits = operation.num_clbits();
+        if !base_gate.is_none() {
+            let instruction = qpy_data
+                .circuit_data
+                .pack(py, &CircuitInstruction::py_new(&base_gate, None, None)?)?;
+            base_gate_raw = serialize(&pack_instruction(
+                &instruction,
+                custom_instructions_hash,
+                new_instructions_list,
+                qpy_data,
+            )?);
+        }
+        Ok(formats::CustomCircuitInstructionDefPack {
+            gate_type,
+            num_qubits,
+            num_clbits,
+            custom_definition: has_definition as u8,
+            num_ctrl_qubits,
+            ctrl_state,
+            name: name.to_string(),
+            data,
+            base_gate_raw,
+        })
+    })
+}
+
+pub fn pack_standalone_vars(
+    circuit_data: &CircuitData,
+    version: u32,
+    standalone_var_indices: &mut HashMap<u128, u16>,
+) -> PyResult<Vec<formats::ExpressionVarDeclarationPack>> {
+    let mut result = Vec::new();
+    let mut index: u16 = 0;
+    let mut uuid: u128 = 0;
+    // input vars
```

Comment:
Consistency suggestion - most of `pack_*` function rely on a `QPYWriteData` parameter. Can we make this something like: 

```suggestion
fn pack_standalone_vars(
    qpy_data: &mut QPYWriteData,
) -> PyResult<Vec<formats::ExpressionVarDeclarationPack>> {
    let mut result = Vec::new();
    let mut index: u16 = 0;
    let mut uuid: u128 = 0;
    // input vars
    let circuit_data = &qpy_data.circuit_data;
    let standalone_var_indices = &mut qpy_data.standalone_var_indices;
    let version = qpy_data.version;
```

Reply:
Yes, will do.